### PR TITLE
chore(iox/compactor): add test cases tracking max_l0_created_at

### PR DESCRIPTION
**Pull request by Joe-Blount**
This PR adds test cases that I believe show a mishandling of max_l0_created_at when overlaps exist between all 3 levels of files.

The new test file takes the approach of describing a base case with a potential vulnerability for mishandling, then adding complicating permutations to that base case trying to trick the compactor into mishandling the potential vulnerability.

More base cases and permutations can likely be added, but I think this is enough to get started.  I'd like to determine if there is agreement on the issue see before spending more time expanding the test.

The naming convention is intended to keep things organized if we have several more base cases with permutations on each.

Helps: https://github.com/influxdata/idpe/issues/17258

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor2/tests/layouts/created_at.rs [hunk 1]
```diff
@@ -0,0 +1,695 @@
...
+#[tokio::test]
+async fn base1_level1_level2_overlaps() {
+    // |------------L1.2------------|                |------------L1.3------------|
+    //                      |--------------L2.1--------------|
+
+    // Base case 1:
+    // This is a base case of two L1s and an L2 overlapping, that will be further complicated
+    // in subsequent test cases.  Assume that each L1 has at least 1 duplicate point with the L2.
+
+    // Note that the L2's max_l0_created_at time is between that of the two L1s.
```
Comment (NGA-TRAN) [inline compactor2/tests/layouts/created_at.rs:28]: The assumption `that the L2's max_l0_created_at time is between that of the two L1s.` is not correct and never happens in our design. If an L2 file overlap in time range with an L1 file, the L2's max_l0_created_at is always older than the one of L1. 

In your case, file L2.1 overlaps with both L1.2 and L1.3. The max_l0_created_at of L2.1 but be older/before the ones of both L1.2 and L1.3

**Review by NGA-TRAN — COMMENTED** [review]
Since the assumption in all of the tests in this PR `that the L2's max_l0_created_at time is between that of the two L1s` cannot happen in our design (see the detail), I wonder  whether you want to modify the assumption first?

Similarly, I have added comments in the ticket https://github.com/influxdata/idpe/issues/17258 that two out of three assumptions  cannot go with each other.  To handle that, we must always compact L0s with all overlapped L1s before compacting L1s  with overlapped L2s.

**Review by NGA-TRAN — CHANGES_REQUESTED** [review]
Have the assumptions to modify first

File: compactor2/tests/layouts/created_at.rs (hunk 1 above)
Comment (alamb) [reply compactor2/tests/layouts/created_at.rs:28]: The idea to add another invariant check to ensure this knowledge is 💯  https://github.com/influxdata/idpe/issues/17258#issuecomment-1464380736

File: compactor2_test_utils/src/lib.rs [hunk 2]
```diff
@@ -749,6 +750,21 @@ fn assert_no_overlap(f1: &ParquetFile, f2: &ParquetFile) {
...
+        && f1.compaction_level < f2.compaction_level
+        && f1.overlaps(f2)
+        && f1.max_l0_created_at < f2.max_l0_created_at
+    {
+        panic!("Found overlapping files with illegal max_l0_created_at order\nf1 = {}\nf2 = {}\n\n{f1:#?}\n\n{f2:#?}",
+               display_format(f1, true),
+               display_format(f2, true),
+        );
+    }
+}
```
Comment (NGA-TRAN) [inline compactor2_test_utils/src/lib.rs:766]: The 2 panic functions look good

File: compactor2/tests/layouts/created_at.rs [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+    // |------------L1.2------------|                |------------L1.3------------|
+    //                      |--------------L2.1--------------|
+
+    // Base case 1:
+    // This is a base case of two L1s and an L2 overlapping, that will be further complicated
+    // in subsequent test cases.  Assume that each L1 has at least 1 duplicate point with the L2.
+
+    // All variations of base case 1 honor the invariant that when overlaps exist between levels,
+    // the lower level file must have a newer max_l0_created_at.  In the above diagram, both L1.2
+    // and L1.3 must have a newer max_l0_created_at than L2.1.
```
Comment (NGA-TRAN) [inline compactor2/tests/layouts/created_at.rs:30]: WIth the way to split and then compact a subset of files, `newer` means `>=` which you do correctly with the panic. I think you want to define newer in this comment, too

File: compactor2/tests/layouts/created_at.rs [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+
+    // Note that the L2's max_l0_created_at time is older than that of the two L1s.
+    // The duplicates in the L1s should prevail.
+
+    // Mishandling the max_lo_created_at in L1.2 is the key vulnerability targeted by this case case.
```
Comment (NGA-TRAN) [inline compactor2/tests/layouts/created_at.rs:35]:
```suggestion
    // Mishandling the max_l0_created_at in L1.2 is the key vulnerability targeted by this case case.
```

File: compactor2/tests/layouts/created_at.rs [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+    // Non-overlapping L1 files but they overlap with L2 files as in the diagram above
+    setup
+        .partition
+        .create_parquet_file(
+            parquet_builder()
+                .with_min_time(1)
+                .with_max_time(5)
+                .with_compaction_level(CompactionLevel::FileNonOverlapped)
+                .with_file_size_bytes(50 * ONE_MB)
+                .with_max_l0_created_at(Time::from_timestamp_nanos(5)), // Older than L2 - That violates an invariant
```
Comment (NGA-TRAN) [inline compactor2/tests/layouts/created_at.rs:259]: 👍

**Pull request merged**